base/src/main/java/org/aya/core/term/Tm.java

```diff
@@ -0,0 +1,219 @@
+// Copyright (c) 2020-2022 Yinsen (Tesla) Zhang.
+// Use of this source code is governed by the MIT license that can be found in the LICENSE.md file.
+package org.aya.core.term;
+
+import kala.collection.immutable.ImmutableMap;
+import kala.collection.immutable.ImmutableSeq;
+import org.aya.concrete.stmt.TopTeleDecl;
+import org.aya.core.Meta;
+import org.aya.core.def.*;
+import org.aya.core.pat.Pat;
+import org.aya.core.repr.AyaShape;
+import org.aya.generic.AyaDocile;
+import org.aya.ref.DefVar;
+import org.aya.ref.LocalVar;
+
+import java.util.function.Function;
+
+/**
+ * A `Tm<T>` is a more general representation of the `Term` AST.
+ * It can be better understood as representing a `Term`-like node,
+ * but with all immediate sub-`Term`s taking values from type `T`.
+ * Here we also provide casting between `Tm<Term>` and `Term` as they are isomorphic.
+ *
+ * @author wsx
+ */
+public sealed interface Tm<T> {
```

Comment:
So `Term` is just `<T extends Tm<T>> Tm<T>`?

Reply:
I think you are right. In more standard terminology `Tm<T>` is the base functor of the inductive type `Term`. And you can get `Term` by getting the fixpoint of `Tm`. But I'm not sure if we need to do this in Java, so currently I just leave the definition of `Term` and `Tm` separate and only link them through the `cast` function.
This definitively leads to some code duplication, do you have any idea to improve?

Reply:
Yes, delete `Term` and refactor relevant operations. Do you have enough spare time to do it?

Reply:
Yes, I can look into refactoring `Term` as well. But I'm not sure how to implement the fixpoint construction in Java?

Reply:
There is a simpler way: make `class Tm<T>` into `class Tm<T extends Tm<T>>`, and after that `Tm<?>` will be equivalent to `Term`.

Reply:
Please bear with me as those Java magic is confusing me. If we define `Tm` as `sealed interface Tm<T extends Tm<T>>`, are we still allowing things like `Tm<String>` even though `String` does not extend `Tm<String>`?

Reply:
No...